server/src/main/java/org/elasticsearch/cluster/coordination/UnsafeBootstrapMasterCommand.java

```diff
@@ -0,0 +1,187 @@
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Objects;
+
+public class UnsafeBootstrapMasterCommand extends EnvironmentAwareCommand {
+
+    private final NamedXContentRegistry namedXContentRegistry;
+
+    static final String WARNING_MSG =
+            "-----------------------------------------------------------------------\n" +
+            "\n" +
+            "    WARNING: Elasticsearch MUST be stopped before running this tool.\n" +
+            "\n" +
+            "You should run this tool only if you've lost the majority of master eligible nodes.\n" +
+            "If you have a backup, restore from the backup instead.\n" +
+            "Running this tool should be the last resort.\n" +
+            "Running this tool may cause arbitrary data loss and may render your cluster completely non functional.\n" +
+            "Do you accept this risk?\n";
+    static final String ABORTED_BY_USER_MSG = "aborted by user";
+    static final String NOT_MASTER_NODE_MSG = "unsafe-bootstrap tool can only be run on master eligible node";
+    static final String FAILED_TO_OBTAIN_NODE_LOCK_MSG = "failed to lock node's directory, is Elasticsearch still running?";
+    static final String NO_NODE_FOLDER_FOUND_MSG = "no node folder is found in data folder(s), node has not been started yet?";
+    static final String NO_NODE_METADATA_FOUND_MSG = "no node meta data is found, node has not been started yet?";
+    static final String NO_MANIFEST_FILE_FOUND_MSG = "no manifest file is found, do you run pre 7.0 Elasticsearch?";
+    static final String GLOBAL_GENERATION_MISSING_MSG = "no metadata is referenced from the manifest file, cluster has never been " +
+            "bootstrapped?";
+    static final String NO_GLOBAL_METADATA_MSG = "failed to find global metadata, metadata corrupted?";
+    static final String EMPTY_LAST_COMMITTED_VOTING_CONFIG_MSG =
+            "last committed voting voting configuration is empty, cluster has never been bootstrapped?";
+    static final String WRITE_METADATA_EXCEPTION_MSG = "exception occurred when writing new metadata to disk";
+    static final String MASTER_NODE_BOOTSTRAPPED_MSG = "Master node was successfully bootstrapped";
+    static final Setting<String> UNSAFE_BOOTSTRAP =
+            ClusterService.USER_DEFINED_META_DATA.getConcreteSetting("cluster.metadata.unsafe-bootstrap");
+
+    UnsafeBootstrapMasterCommand() {
+        super("Unsafely bootstraps the master node if the majority of master eligible nodes is lost");
+        namedXContentRegistry = new NamedXContentRegistry(ClusterModule.getNamedXWriteables());
+    }
+
+    @Override
+    protected void execute(Terminal terminal, OptionSet options, Environment env) throws Exception {
+        // We don't use classical "private static final Logger", because log4j2 initialization happens
+        // after UnsafeBootstrapMasterCommand instance creation.
```

Comment:
Hmm, we make a `private static final Logger logger` in `RemoveCorruptedShardDataCommand`, does that not work?

Also, does this logger have the same configuration as loggers in Elasticsearch proper, i.e., it writes to the Elasticsearch log by default? If so, I think we should log more information about this tool having been run in that log.

Reply:
Ok, I have figured out how the logging works:

1. Most of the commands, for example, `LoggingAwareCommand` are passing `CommandLoggingConfigurator::configureLoggingWithoutConfig` callback to `Command` constructor as `beforeMain` argument. This is the function that will be executed when `Command.main` is called.
2. From CommandLoggingConfigurator::configureLoggingWithoutConfig javadoc `Configure logging without reading a log4j2.properties file, effectively configuring the status logger and all loggers to the console.` Basically it means that logs won't be appended to elasticsearch log files.
3. If you run `ShardToolCli` indeed you get `ERROR StatusLogger No Log4j 2 configuration file found. Using default configuration`. The reason for this is because logger initialization happens when log4j2 classes are loaded. Java Runtime loads classes when it first encounters this class, in case of `ShardToolCli` it will load `RemoveCorruptedShardDataCommand` and transitively `Logger` class, because there is static reference to `Logger` class in `RemoveCorruptedShardDataCommand`. Since pre-main is not yet executed, logging is not configured.
4. It turned out that `NodeToolCli` still suffers from the same warning, because there is `ClusterService` (which in turn staticly references `Logger`) static reference in `UnsafeBootstrapMasterCommand`. The reason why I've not noticied it earlier, because I've added this setting last moment, after I've checked how binaries are working.
5. For now I've made the following fix https://github.com/elastic/elasticsearch/pull/37696/commits/e23f59af291f0bc60af9b2a7bd252815950632b9
6. I suggest to revisit how logging works for commands in general in the follow-up PR. It seems that `beforeMain` could be completely removed and replaced by logging initialization in the command constructor. But this change touches a lot of file and I prefer separate PR.